go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,551 @@
+	cluster         *OvnClusterController
+	ovnController   *ovn.Controller
+	nodeName        string
+	manageDBServers bool
+	isLeader        bool
+	leaderElector   *leaderelection.LeaderElector
+}
+
+// NewHAMasterController creates a new HA Master controller
+func NewHAMasterController(kubeClient kubernetes.Interface, cluster *OvnClusterController,
+	nodeName string, manageDBServers bool) *OvnHAMasterController {
+	ovnController := ovn.NewOvnController(kubeClient, cluster.watchFactory)
+	return &OvnHAMasterController{
+		kubeClient:      kubeClient,
+		cluster:         cluster,
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+		isLeader:        false,
+		leaderElector:   nil,
+	}
+}
+
+// StartHAMasterCluster runs the replication controller
+func (hacluster *OvnHAMasterController) StartHAMasterCluster() error {
+	if hacluster.manageDBServers {
+		// Always demote the OVN DBs to backmode mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacluster.DemoteOVNDbs(invalidIPAddress, ovnNorthDbPort, ovnSouthDbPort)
+		if err != nil {
+			// If we are not able to communicate to the OVN ovsdb-servers,
+			// then it is better to return than continue.
+			// cmd/ovnkube.go will panic if this function returns error.
+			return err
+		}
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacluster.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacluster.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	HAClusterOnStoppedLeading := func() {
+		//This node was leader and it lost the election.
+		// Whenever the node transitions from leader to follower,
+		// we need to handle the transition properly like clearing
+		// the cache. It is better to exit for now.
+		// kube will restart and this will become a follower.
+		if hacluster.manageDBServers {
+			// Stop ovn-northd and then exit.
+			_, _, _ = util.RunOVNNorthAppCtl("exit")
+		}
+		logrus.Infof("I (" + hacluster.nodeName + ") am no longer a leader. Exiting")
+		os.Exit(1)
+	}
+
+	HAClusterNewLeader := func(nodeName string) {
+		logrus.Infof(nodeName + " is the new leader")
+		wasLeader := hacluster.isLeader
+
+		if hacluster.nodeName == nodeName {
+			logrus.Infof(" I (" + hacluster.nodeName + ") won the election. In active mode")
+			err = hacluster.ConfigureAsActive(nodeName)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				if hacluster.manageDBServers {
+					// Stop ovn-northd before panicing.
+					_, _, _ = util.RunOVNNorthAppCtl("exit")
+				}
+				panic(err.Error())
+			}
+			hacluster.isLeader = true
+		} else {
+			logrus.Infof(" I (" + hacluster.nodeName + ") lost the election. In Standby mode")
+			if wasLeader {
+				HAClusterOnStoppedLeading()
```

Comment:
@numansiddique noww that HAClusterOnStoppedLeading() calls os.Exit(1), for clarity should we do:

```
if hacluster.nodeName == nodeName {
   <configure as leader>
} else if wasLeader {
   HAClusterOnStoppedLeading()
   // should not be reached
   panic()
} else {
   <configure as standby>
}
```

Reply:
Done